### ISSUE TITLE
Abacus 2 Gpc/h inference

### ISSUE THREAD

**Issue by maho3** [maho3/ltu-gobig#23]

**Comment by maho3** [comment]
## Coverage
![image](https://github.com/user-attachments/assets/ab277081-9127-48fe-b3d0-dd0346ccd684)

## $k_{\rm min} = 0.2$ inference
![image](https://github.com/user-attachments/assets/cc2ce9be-fef7-4fde-afb9-9bbd46d7b405)
![image](https://github.com/user-attachments/assets/60fc811f-caf0-49e2-b82e-4270f46e576d)
![image](https://github.com/user-attachments/assets/48ea29b3-d49d-46a5-a302-c2692e1060bc)

**Comment by maho3** [comment]
Pk0+Pk2 kmax=0.3
![Image](https://github.com/user-attachments/assets/fb98c415-5144-4043-828e-b355d6cce81f)

Pk0+Pk2+Qk0+Qk2 kmax=0.3
![Image](https://github.com/user-attachments/assets/25f5edb5-2cc3-482e-9af1-cc507452fd0e)

**Comment by maho3** [comment]
Training on abacuslike/halos, testing on abacus/halos. Using zPk024+zQk0. I'm getting lots of out-of-distribution points and I don't know why.

![Image](https://github.com/user-attachments/assets/53ea3707-f059-4869-92c6-b90a7574e505)

**Comment by CompiledAtBirth** [comment]
_Investigation regarding Abacus halo Finder_
Number of halos inside 2Gpc/h cube comoving boxes from the [Abacus Emulator Grid simulations](https://abacussummit.readthedocs.io/en/latest/cosmologies.html), at redshift $z=0.5$.
![Image](https://github.com/user-attachments/assets/2a9fe6ae-8722-4c3c-8b54-6ecea3fd3fc3)

**Comment by CompiledAtBirth** [comment]
Plots of Qk0 from the cmass pipeline and from Abacus simulations. Colors are for vizualization purposes.
The **orange line** in the cmass panels correspond to one Abacus simulation example.

![Image](https://github.com/user-attachments/assets/751a2308-54d9-4c81-b9c9-59531ab4d732)

**Comment by CompiledAtBirth** [comment]
![Image](https://github.com/user-attachments/assets/6c0fc250-74a3-4dd2-bcb5-fb0415399c8d)

**Comment by CompiledAtBirth** [comment]
same cosmologies as above
![Image](https://github.com/user-attachments/assets/64196d81-5d58-4027-9fa2-c08fb590a896)

**Comment by CompiledAtBirth** [comment]
![Image](https://github.com/user-attachments/assets/0dfa71c9-45c5-49bd-97b6-0c76f1204831)

**Comment by maho3** [comment]
Here I'm comparing the summary statistics of Abacus halos and their cosmology-matched CHARM counterparts:

![Image](https://github.com/user-attachments/assets/0bbf065e-74ec-4036-8638-a96dd12525af)

**Comment by maho3** [comment]
## Relative constraining power on a  simbig-esque SGC cut

**Comment by maho3** [comment]
Training on CHARM with Zheng07 + Assem/Velocity Bias. Testing on Abacus with AbacusHOD

![Image](https://github.com/user-attachments/assets/c513c9d8-7780-494d-b147-fef15a187667)

![Image](https://github.com/user-attachments/assets/d1d9d423-ec19-45dc-adc0-45801a95827d)

**Comment by maho3** [comment]
We don't have as much constraining power as we expect, in comparison to [SIMBIG's Bk analysis](https://arxiv.org/pdf/2310.15243). @shivampcosmo proposed that this is due to the fact that our logMmin is unconstrained by a wide prior, and is very degenerate with sigma8, whereas SIMBIG fixes logMmin based on the nbar of the halo field.

As a quick test, we add nbar as another feature with which to train on, to try to break the degeneracy between sigma8 and logMmin.

## Abacuslike galaxies, self-consistent tests, with/without nbar
### zPk0

![Image](https://github.com/user-attachments/assets/8a1eb574-a463-4134-8731-921a0e044d59)

![Image](https://github.com/user-attachments/assets/43a8f0a3-7c05-4a66-a6ca-db9167381cd2)

### zPk024+Qk0

![Image](https://github.com/user-attachments/assets/c418bc2f-2ae2-42d5-a69b-3f1f8b8ecceb)

![Image](https://github.com/user-attachments/assets/29d3be35-9744-4e16-b855-e6d015661795)

**Comment by maho3** [comment]
@florpi suggested that the constraining power could be improved with NLE. I was only able to test this with the zPk0 statistics (the others need compression), but here is a self-consistent prediction showing the constraining power :

![Image](https://github.com/user-attachments/assets/203ad261-edad-41e4-8a8e-f843dbacbc6f)

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Omega_m</th>
      <th>Omega_b</th>
      <th>n_s</th>
      <th>h</th>
      <th>sigma8</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>1000.000000</td>
      <td>1000.000000</td>
      <td>1000.000000</td>
      <td>1000.000000</td>
      <td>1000.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>0.440132</td>
      <td>0.046428</td>
      <td>0.681116</td>
      <td>1.081692</td>
      <td>0.840546</td>
    </tr>
    <tr>
      <th>std</th>
      <td>0.030986</td>
      <td>0.008419</td>
      <td>0.082746</td>
      <td>0.073309</td>
      <td>0.094591</td>
    </tr>
    <tr>
      <th>min</th>
      <td>0.330555</td>
      <td>0.030053</td>
      <td>0.503247</td>
      <td>0.837290</td>
      <td>0.609483</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>0.420513</td>
      <td>0.040124</td>
      <td>0.622516</td>
      <td>1.028722</td>
      <td>0.771306</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>0.442173</td>
      <td>0.045871</td>
      <td>0.675934</td>
      <td>1.084800</td>
      <td>0.854388</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>0.462321</td>
      <td>0.052398</td>
      <td>0.737712</td>
      <td>1.143761</td>
      <td>0.916320</td>
    </tr>
    <tr>
      <th>max</th>
      <td>0.499728</td>
      <td>0.069965</td>
      <td>0.897294</td>
      <td>1.199809</td>
      <td>0.999245</td>
    </tr>
  </tbody>
</table>
</div>

**Comment by maho3** [comment]
## Now training on CHARM, testing on Abacus, but with Optuna hyperparameter tuning

![Image](https://github.com/user-attachments/assets/726f9a24-dba5-4945-9f13-3caa3c6fa896)
![Image](https://github.com/user-attachments/assets/159c586a-0dfa-4ead-b070-3bac74e8fe55)
![Image](https://github.com/user-attachments/assets/b1e8abce-3093-4d85-a19f-23738ed29126)

**Comment by maho3** [comment]
## Optuna hyperparameter-tuned, train CHARM, test Abacus
We've adapted our training procedure to now use Optuna. (See https://github.com/maho3/ltu-cmass/pull/115 ) 

We've also run more HODs per abacuslike simulation. There's now 5 HOD's per simulation. All results below are for galaxies, with RSDs, unless stated.

We gain in constraining power. Interestingly, we are unbiased across most summaries except the full bispectrum. This could perhaps be a result of not sufficient training data, as the Qk0 vector is much longer than Pk024.

![Image](https://github.com/user-attachments/assets/bcbe8848-cf3b-42a0-8468-54ab19d19f6d)

![Image](https://github.com/user-attachments/assets/77578132-7c2b-49a0-aaea-49556a35cd45)

![Image](https://github.com/user-attachments/assets/e215620c-f9b3-4438-ba81-a26d6f4e36b6)

![Image](https://github.com/user-attachments/assets/20a2625d-f6cf-4094-b3c1-38105d89c07c)

This also occurs when we exclude nbar

![Image](https://github.com/user-attachments/assets/c32d2651-5bd7-44a8-935b-c40c6f1a32ff)

However, it goes away when we exclude RSDs.

![Image](https://github.com/user-attachments/assets/37f6696e-6d2e-4bac-875d-72ba2bdb0d21)

**Comment by maho3** [comment]
## Optuna train CHARM, test Abacus, Squeezed and Subsampled Bispectrum
Squeezed Bispectrum
![Image](https://github.com/user-attachments/assets/0d7cd468-3d32-40f1-af3a-31832d7a1d9f)

![Image](https://github.com/user-attachments/assets/99916d1d-9584-4968-a442-f390952185a7)

Subsampled bispectrum
![Image](https://github.com/user-attachments/assets/2fcddbbe-f155-48e1-b8b2-1ceebb34fba1)

![Image](https://github.com/user-attachments/assets/639fe981-62be-49ef-ac91-4b8897f971f9)

**Comment by maho3** [comment]
# Updated training script, with optuna, testing on Abacus
Following [this PR](https://github.com/maho3/ltu-cmass/pull/120), we've added some changes:
* Added isoceles bispectrum as a possible feature
* Added learning rate scheduler parameters to optuna hyperparameter tuning
* Added global mean/std normalization to ltu-ili

The goal was for these to better manage the training process, and reduce biases in the predictions

## Training on CHARM galaxies with RSD, testing on Abacus

* Power spectra
![Image](https://github.com/user-attachments/assets/abe965d5-5389-4548-b218-e14741322ab5)
* Equilateral Bk
![Image](https://github.com/user-attachments/assets/2a60ae83-ca60-4012-8f11-bcf684ea4aaf)
* Squeezed Bk
![Image](https://github.com/user-attachments/assets/476ced93-629a-4321-bbfb-dea2883bc316)
* Isoceles Bk
![Image](https://github.com/user-attachments/assets/e6e9b593-c7ca-4d23-9cc0-2e07dfc425b7)
* Full Qk
![Image](https://github.com/user-attachments/assets/40c9ffae-63e9-4fee-b609-75b72b6aaced)

Takeaways: 
* We get a little bit more constraining power with the recent changes
* The underprediction of Omega_m starts to take effect in the Isoceles and Full Qk. These biases are not resolved.

## Another possible explanation for what could be going wrong.
In all of the predictions on Abacus, we seem to overpredict n_s.
![Image](https://github.com/user-attachments/assets/577919d5-2513-4102-97c9-b02f35852624)

And n_s is slightly degenerate with Omega_m, in the correct direction such that an overprediction of n_s could lead to an underprediction of Omega_m, and vice-versa. Perhaps this is due to degenerate effects of n_s and neutrinos? (not likely the only reason)
![Image](https://github.com/user-attachments/assets/49d81bdc-1fd8-467b-8539-9ad3bd4be853)

In theory, if we were to fix n_s at its 'true value,' the this would bring our Omega_m predictions into alignment. In the bottom plot, I've only taken exactly same example as the previous figure^, but sliced samples from the posterior 'around' the true n_s (+/- 0.03).
![Image](https://github.com/user-attachments/assets/9e612b81-b794-4a01-a995-491f69c0acec)
This seems to bring Omega_m slightly closer to the true value, but does not fix our problem outright (within 1 sigma for example).

**Comment by maho3** [comment]
## Uniform+Gaussian Noising scheme, Jeffrey's prior, evaluating on Abacus

Here, I'm training on CHARM simulations with the new noising scheme (Uniform+Gaussian) where the Gaussian's sigma_radial and sigma_transverse are set by a Jeffrey's prior where log(sigma) is uniform. Then, I'm testing on real Abacus halos, where I don't add any noise (i.e. sigma_radial=sigma_transverse=0). All are done for sgc_lightcone summaries, with z-dependent HOD.

* Power spectra
![Image](https://github.com/user-attachments/assets/1a26d185-082c-4b19-bbec-c10ec26f7d32)
* Squeezed bispectrum
![Image](https://github.com/user-attachments/assets/d131806a-6189-43fb-8031-5c26a1b32c91)
* Iscosceles bispectrum
![Image](https://github.com/user-attachments/assets/4d79fbc2-fc2d-4ca3-be20-ec9591bbe2d3)
* Full bispectrum
![Image](https://github.com/user-attachments/assets/a82a182b-49f5-4a72-b6a3-412f32cd588a)

Takeaways:
* This seems to remove the bias on Omega_m!
* As in https://github.com/maho3/ltu-gobig/issues/22#issuecomment-3021778025, the sigma8 constraints are all over the place, and not very tight. I suspect this is because of either (1) sigma_radial is degenerate with sigma8 and causing problems or (2) we don't have enough simulations to efficiently constrain our models.

**Comment by maho3** [comment]
## Uniform+Gaussian Noising scheme + NGP meshing, Jeffrey's prior, evaluating on Abacus

Same thing as previous, but now we're using NGP to mesh the halos (to avoid substructure)

* Pk024
![Image](https://github.com/user-attachments/assets/311d5d88-a534-4954-93b1-5a4bf959275c)

* Isosceles bispectrum
![Image](https://github.com/user-attachments/assets/8e2ddb02-eb59-458e-ac84-af13074d4db1)

* Qk0
![Image](https://github.com/user-attachments/assets/899d38bf-fa3e-4e2b-a8be-b1c59fbc8c40)

Note there's only a few points, because I ran one HOD per simulation and only selected those with galaxy number densities near that of CMASS.

**Comment by suicee** [comment]
## Exponential Noise evaluating on Abacus

* Pk0

![Image](https://github.com/user-attachments/assets/0fb9f42f-a483-47ac-ad22-9f638c160149)

* Pk024

![Image](https://github.com/user-attachments/assets/f3362865-1593-4f4d-a7a7-d0a9a5c6685b)

* Pk024+ Equilateral bispectrum

![Image](https://github.com/user-attachments/assets/f5a77f71-0a1c-49c7-b848-979b420e3c70)

**Comment by maho3** [comment]
We fixed the n_s issue in https://github.com/maho3/ltu-gobig/issues/36.

Putting here some plots that I generated for the ILI meeting. At some point, we should come back and test this for all summaries.

## Uniform+Gaussian Noising scheme + NGP meshing, Jeffrey's prior, evaluating on Abacus w/ 4 Mpc/h noise
**Train on**: 2 Gpc/h CHARM+Zheng07ex with Uniform+Gaussian noise
**Noise prior**: Reciprocal(0.1, 4.5) Mpc/h
**Tracer**: sgc_lightcone
**Test on**: Abacus+Zheng07ex with 4 Mpc/h Gaussian noise

* Full Pk024+Qk0
<img width="850" alt="Image" src="https://github.com/user-attachments/assets/387db7b9-8a8b-4583-a7dd-b9165f1476c8" />

<img width="540" alt="Image" src="https://github.com/user-attachments/assets/c54c4e52-bd63-4a62-af90-c43e1f5b6dc5" />

## TODO
* Look into why sigma8 is so noisy

**Comment by maho3** [comment]
Same as above^, but for MTNG

## Uniform+Gaussian Noising scheme + NGP meshing, Jeffrey's prior, evaluating on MTNG lightcone w/ 4 Mpc/h noise
**Train on**: 2 Gpc/h CHARM+Zheng07ex with Uniform+Gaussian noise
**Noise prior**: Reciprocal(0.1, 4.5) Mpc/h
**Tracer**: sgc_lightcone
**Test on**: MTNG lightcone with 4 Mpc/h Gaussian noise

* Full Pk024+Qk0

<img width="540" alt="Image" src="https://github.com/user-attachments/assets/c8c36e4f-eb9b-4aff-b25b-4cb360f153bf" />
